File: backend/jasprchain/crypto/keys.py
"""Ed25519 Key Management for JasprChain Wallets
Mirrors Rust: ed25519_dalek crate
"""
import nacl.signing
import nacl.encoding
import hashlib
import secrets
from dataclasses import dataclass
from typing import Optional
import base64


@dataclass
class Ed25519KeyPair:
    """Ed25519 keypair for wallet operations"""
    private_key: bytes
    public_key: bytes
    address: str
    
    def sign(self, message: bytes) -> bytes:
        """Sign a message with the private key"""
        signing_key = nacl.signing.SigningKey(self.private_key)
        signed = signing_key.sign(message)
        return signed.signature
    
    def verify(self, message: bytes, signature: bytes) -> bool:
        """Verify a signature"""
        try:
            verify_key = nacl.signing.VerifyKey(self.public_key)
            verify_key.verify(message, signature)
            return True
        except nacl.exceptions.BadSignature:
            return False
    
    def to_dict(self) -> dict:
        return {
            'public_key': base64.b64encode(self.public_key).decode(),
            'address': self.address
        }


def generate_wallet(seed: Optional[bytes] = None) -> Ed25519KeyPair:
    """Generate a new wallet keypair"""
    if seed:
        # Deterministic key generation from seed
        seed_hash = hashlib.sha256(seed).digest()
        signing_key = nacl.signing.SigningKey(seed_hash)
    else:
        # Random key generation
        signing_key = nacl.signing.SigningKey.generate()
    
    verify_key = signing_key.verify_key
    
    # Create address from public key (first 20 bytes of hash)
    address_hash = hashlib.sha256(bytes(verify_key)).hexdigest()[:40]
    address = f"jaspr1{address_hash}"
    
    return Ed25519KeyPair(
        private_key=bytes(signing_key),
        public_key=bytes(verify_key),
        address=address
    )


def address_from_public_key(public_key: bytes) -> str:
    """Derive address from public key"""
    address_hash = hashlib.sha256(public_key).hexdigest()[:40]
    return f"jaspr1{address_hash}"


def verify_signature(public_key: bytes, message: bytes, signature: bytes) -> bool:
    """Standalone signature verification"""
    try:
        verify_key = nacl.signing.VerifyKey(public_key)
        verify_key.verify(message, signature)
        return True
    except:
        return False

File: rust-core/jasprchain/src/crypto/keys.rs
//! Ed25519 key management for wallets

use ed25519_dalek::{SigningKey, VerifyingKey, Signature, Signer, Verifier};
use sha2::{Sha256, Digest};
use rand::rngs::OsRng;

/// Ed25519 keypair for wallet operations
pub struct Ed25519KeyPair {
    signing_key: SigningKey,
    verifying_key: VerifyingKey,
    pub address: String,
}

impl Ed25519KeyPair {
    /// Generate new keypair
    pub fn generate() -> Self {
        let signing_key = SigningKey::generate(&mut OsRng);
        let verifying_key = signing_key.verifying_key();
        
        let address = Self::derive_address(&verifying_key);
        
        Self {
            signing_key,
            verifying_key,
            address,
        }
    }
    
    /// Generate from seed
    pub fn from_seed(seed: &[u8; 32]) -> Self {
        let signing_key = SigningKey::from_bytes(seed);
        let verifying_key = signing_key.verifying_key();
        
        let address = Self::derive_address(&verifying_key);
        
        Self {
            signing_key,
            verifying_key,
            address,
        }
    }
    
    fn derive_address(verifying_key: &VerifyingKey) -> String {
        let mut hasher = Sha256::new();
        hasher.update(verifying_key.as_bytes());
        let hash = hex::encode(hasher.finalize());
        format!("jaspr1{}", &hash[..40])
    }
    
    /// Sign a message
    pub fn sign(&self, message: &[u8]) -> Vec<u8> {
        let signature: Signature = self.signing_key.sign(message);
        signature.to_bytes().to_vec()
    }
    
    /// Verify a signature
    pub fn verify(&self, message: &[u8], signature: &[u8]) -> bool {
        if signature.len() != 64 {
            return false;
        }
        let sig_bytes: [u8; 64] = signature.try_into().unwrap();
        let signature = Signature::from_bytes(&sig_bytes);
        self.verifying_key.verify(message, &signature).is_ok()
    }
    
    pub fn public_key_bytes(&self) -> Vec<u8> {
        self.verifying_key.as_bytes().to_vec()
    }
}

/// Generate a new wallet
pub fn generate_wallet() -> Ed25519KeyPair {
    Ed25519KeyPair::generate()
}
